### PR TITLE
Swerve drive

### PR DESCRIPTION
**Pull request by adamv17**
This pull request was made to create a swerve drive. The swerve drive has 4 wheels with 2 independent motors for drive and rotation. The mathematics of swerve is in the SwerveDrive Subsystem while the SwerveModule commands the motors of each wheel.

### CODE REVIEW

File: src/main/java/robot/Constants.java [hunk 1]
```diff
@@ -18,6 +19,21 @@
         public static final int MAX_CURRENT = 35;
     }
 
+    public static class SwerveDrive {
+        public static final double ROBOT_LENGTH = 0;
+        public static final double ROBOT_WIDTH = 0;
+        public static final double SPEED_MULTIPLIER = 0.7;
+        public static final double ROTATION_MULTIPLIER = 2 * Math.PI;
```
Comment (orel-david) [inline src/main/java/robot/Constants.java]: What does this constants mean?

File: src/main/java/robot/Constants.java [hunk 2]
```diff
@@ -18,6 +19,21 @@
         public static final int MAX_CURRENT = 35;
     }
 
+    public static class SwerveDrive {
+        public static final double ROBOT_LENGTH = 0;
+        public static final double ROBOT_WIDTH = 0;
```
Comment (orel-david) [inline src/main/java/robot/Constants.java]: Please mention that they aren't the actual constants

File: src/main/java/robot/OI.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        return joystick.getX();
+    }
+
+    public static double getJoystickY() {
+        return joystick.getY();
+    }
+
+    public static double getJoystickZ() {
+        return joystick.getZ();
+    }
```
Comment (orel-david) [inline src/main/java/robot/OI.java]: Add documentation for the methods

File: src/main/java/robot/Ports.java [hunk 4]
```diff
@@ -9,4 +9,16 @@
...
+
+    public static class SwerveDrive {
+        public static final int frontRightDrive = 11;
+        public static final int frontRightAngle = 12;
+        public static final int frontLeftDrive = 21;
+        public static final int frontLeftAngle = 22;
+        public static final int backRightDrive = 31;
+        public static final int backRightAngle = 32;
+        public static final int backLeftDrive = 41;
+        public static final int backLeftAngle = 42;
```
Comment (orel-david) [inline src/main/java/robot/Ports.java]: I suggest that you give each side a serial number for example front right would be 1 and front left would be 2. In addition to that please mention that they are motors.

File: src/main/java/robot/Utils.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+     * @return the remainder of the division
+     */
+    public static double floorMod(double value, double mod) {
+        value %= mod;
+        value += mod;
+        value %= mod;
+        return value;
+    }
+
+    public static double[] cartesianToPolar(double x, double y) {
```
Comment (orel-david) [inline src/main/java/robot/Utils.java:26]: Please document this method

File: src/main/java/robot/Utils.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        for (int i = 0; i < m.length; i++) {
+            sum = 0;
+            for (int j = 0; j < v.length; j++) {
+                sum += m[i][j] * v[j];
+            }
+            out[i] = sum;
+        }
+
+        return out;
+    }
```
Comment (orel-david) [inline src/main/java/robot/Utils.java]: Please document this method as well

File: src/main/java/robot/subsystems/drivetrain/SwerveDrive.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            if (i % 2 != 0) {
+                M[i][0] = 1;
+                M[i][1] = 0;
+                M[i][2] = r * signX[i/2];
+            } else {
+                M[i][0] = 0;
+                M[i][1] = 1;
+                M[i][2] = r * signY[i/2];
+            }
+        }
```
Comment (orel-david) [inline src/main/java/robot/subsystems/drivetrain/SwerveDrive.java]: Please document this method and mention the control book for the math

File: src/main/java/robot/subsystems/drivetrain/SwerveDrive.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        double[] signX = {1, 1, -1, -1};
+        double[] signY = {-1, 1, 1, -1};
+
+        double[][] M = new double[8][3];
+
+        for (int i = 0; i < 8; i++) {
+            if (i % 2 != 0) {
+                M[i][0] = 1;
+                M[i][1] = 0;
+                M[i][2] = r * signX[i/2];
```
Comment (orel-david) [inline src/main/java/robot/subsystems/drivetrain/SwerveDrive.java]: Notice that you multiply by the X component of the radius which is `ROBOT_WIDTH/2`. Also please notice that `i` is an odd   number so you need to round the index of `i/2`

File: src/main/java/robot/subsystems/drivetrain/SwerveDrive.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+            } else {
+                M[i][0] = 0;
+                M[i][1] = 1;
+                M[i][2] = r * signY[i/2];
```
Comment (orel-david) [inline src/main/java/robot/subsystems/drivetrain/SwerveDrive.java]: Notice that you multiply by the Y component of the radius which is `ROBOT_HEIGHT/2`.

File: src/main/java/robot/subsystems/drivetrain/SwerveDrive.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        }
+
+        double[] wheelVelocities = Utils.matrixVectorMult(M, robotHeading);
+
+        return wheelVelocities;
+
+    }
+
+
+    public void holonomicDrive(double forward, double strafe, double rotation) {
```
Comment (orel-david) [inline src/main/java/robot/subsystems/drivetrain/SwerveDrive.java]: Please document this method

File: src/main/java/robot/subsystems/drivetrain/SwerveDrive.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        forward *= SPEED_MULTIPLIER;
+        strafe *= SPEED_MULTIPLIER;
+        rotation *= ROTATION_MULTIPLIER;
+
+        if (isFieldOriented) {
+            double robotAngle = gyro.getAngle();
+            double tmp = forward * Math.cos(robotAngle) + strafe * Math.sin(robotAngle);
+            strafe = (-1) * forward * Math.sin(robotAngle) + strafe * Math.cos(robotAngle);
+            forward = tmp;
```
Comment (orel-david) [inline src/main/java/robot/subsystems/drivetrain/SwerveDrive.java]: Please add documentation for the math

File: src/main/java/robot/subsystems/drivetrain/SwerveDrive.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        }
+
+        double[] velocities = calculateWheelVelocities(forward, strafe, rotation);
+        double[] polar;
+        double[][] controls = new double[4][2];
+
+        for (int i = 0; i < velocities.length; i += 2) {
+            polar = Utils.cartesianToPolar(velocities[i], velocities[i + 1]);
+            controls[i/2][0] = polar[0];
+            controls[i/2][1] = polar[1];
```
Comment (orel-david) [inline src/main/java/robot/subsystems/drivetrain/SwerveDrive.java]: Notice that `i/2` might be a double and this is an index. also why do you need to divide `i`? after all this loop would iterate 4 times due to the fact that you incremant it by 2

File: src/main/java/robot/subsystems/drivetrain/SwerveModule.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+        angleMotor.config_kI(wheel, KI, TALON_TIMEOUT);
+        angleMotor.config_kD(wheel, KD, TALON_TIMEOUT);
+        angleMotor.config_kF(wheel, KF, TALON_TIMEOUT);
+
+
+        this.driveMotor = driveMotor;
+        this.angleMotor = angleMotor;
+    }
+
+    public double getSpeed() {
```
Comment (orel-david) [inline src/main/java/robot/subsystems/drivetrain/SwerveModule.java:57]: Document this method

File: src/main/java/robot/subsystems/drivetrain/SwerveModule.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        return unit.toVelocity(driveMotor.getSelectedSensorVelocity());
+    }
+
+    public double getAngle() {
```
Comment (orel-david) [inline src/main/java/robot/subsystems/drivetrain/SwerveModule.java:64]: Document this method as well

File: src/main/java/robot/subsystems/drivetrain/SwerveModule.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        return unit.toUnits(angleMotor.getSelectedSensorPosition());
+    }
+
+    public void setSpeed(double speed) {
+        driveMotor.set(ControlMode.PercentOutput, unit.toTicks100ms(speed));
+    }
+
+    public void setTargetAngle(double angle) {
```
Comment (orel-david) [inline src/main/java/robot/subsystems/drivetrain/SwerveModule.java]: Please explain the logic of this method

File: src/main/java/robot/subsystems/drivetrain/UnitModel.java [hunk 16]
```diff
@@ -0,0 +1,26 @@
+package robot.subsystems.drivetrain;
+
+public class UnitModel {
```
Comment (orel-david) [inline src/main/java/robot/subsystems/drivetrain/UnitModel.java:3]: Please add documentation for the methods

File: src/main/java/robot/subsystems/drivetrain/commands/HolonomicDrive.java [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+        double rotation = OI.getJoystickZ();
+
+        forward = joystickDeadband(forward);
+        strafe = joystickDeadband(strafe);
+        rotation = joystickDeadband(rotation);
+
+        swerveDrive.holonomicDrive(forward, strafe, rotation);
+    }
+
+    private double joystickDeadband(double val) {
```
Comment (orel-david) [inline src/main/java/robot/subsystems/drivetrain/commands/HolonomicDrive.java:38]: Please add documentation

**Review by orel-david — CHANGES_REQUESTED** [review]
In general you did a great job, but some of the methods lack some documentation.

File: src/main/java/robot/subsystems/drivetrain/SwerveDrive.java (hunk 8 above)
Comment (adamv17) [reply src/main/java/robot/subsystems/drivetrain/SwerveDrive.java]: i is an int so it will round by itself, i.e (int) 7 / (int) 2 = (int) 3

File: src/main/java/robot/subsystems/drivetrain/SwerveDrive.java (hunk 12 above)
Comment (adamv17) [reply src/main/java/robot/subsystems/drivetrain/SwerveDrive.java]: The 2D array controls has a size of 4 by 2 while the array called velocities has 8 indices. Therefore for each pair of velocities there is one row of controls hence i/2.

File: src/main/java/robot/subsystems/drivetrain/SwerveDrive.java (hunk 9 above)
Comment (adamv17) [reply src/main/java/robot/subsystems/drivetrain/SwerveDrive.java]: Can you rephrase? I don't understand your comment

Comment (adamv17) [reply src/main/java/robot/subsystems/drivetrain/SwerveDrive.java]: I think I get it now

File: src/main/java/robot/Constants.java [hunk 18]
```diff
@@ -9,15 +9,41 @@
...
         public static final double MAX_VEL = 3;// in m/s
         public static final double TIME_STEP = 0.02;
         public static final double MAX_ACCELERATION = 0.4;// in m/s^2 (currently not the correct number)
         public static final int MAX_CURRENT = 35;
     }
 
+    public static class SwerveDrive {
+        // TODO: not actual measurements and constants
+        public static final double ROBOT_LENGTH = 0;
+        public static final double ROBOT_WIDTH = 0;
```
Comment (orel-david) [inline src/main/java/robot/Constants.java]: Mention the units of this measurements

File: src/main/java/robot/Constants.java [hunk 19]
```diff
@@ -9,15 +9,41 @@
...
+    public static final int TALON_TIMEOUT = 10;
 
     public static class Drivetrain {
         public static final double TICKS_PER_METER = 256 / (4 * 0.0254 * Math.PI);
+        public static final double TICKS_PER_DEGREE = 1; // TODO: change to real value
+
         public static final double MAX_VEL = 3;// in m/s
         public static final double TIME_STEP = 0.02;
         public static final double MAX_ACCELERATION = 0.4;// in m/s^2 (currently not the correct number)
         public static final int MAX_CURRENT = 35;
```
Comment (orel-david) [inline src/main/java/robot/Constants.java]: Mention units

**Review by orel-david — CHANGES_REQUESTED** [review]

File: src/main/java/robot/subsystems/drivetrain/SwerveDrive.java [hunk 20]
```diff
@@ -105,6 +105,19 @@ public void holonomicDrive(double forward, double strafe, double rotation) {
...
+    /**
+     * set the angle of the wheels on the robot to lock the robot in place
+     */
+    public void lock() {
+        // calculates the lock angle of the wheels
+        double lockAngle = Math.PI / 2 - Math.tan(ROBOT_WIDTH / ROBOT_LENGTH);
+
+        for (int i = 0; i < 4; i++) {
+            swerveModules[i].setSpeed(0);
+            swerveModules[i].setTargetAngle(lockAngle);
```
Comment (orel-david) [inline src/main/java/robot/subsystems/drivetrain/SwerveDrive.java]: You should give every wheel different angle

File: src/main/java/robot/Main.java [hunk 21]
```diff
@@ -1,29 +0,0 @@
...
-package robot;
-
-import edu.wpi.first.wpilibj.RobotBase;
-
-/**
- * Do NOT add any static variables to this class, or any initialization at all.
- * Unless you know what you are doing, do not modify this file except to
- * change the parameter class to the startRobot call.
- */
-public final class Main {
```
Comment (katzuv) [inline src/main/java/robot/Main.java]: Why did you delete this file?

File: src/main/java/robot/Ports.java [hunk 22]
```diff
@@ -9,4 +9,23 @@
     public static class ExampleSubsystem1 {
         //public static int TALON_PORT = 1;
     }
+
+    public static class SwerveDrive {
+        // front right
```
Comment (katzuv) [inline src/main/java/robot/Ports.java:14]: AFAIR, Galaxia's convention was left-first, front-first.

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/robot/Ports.java (hunk 22 above)
Comment (orel-david) [reply src/main/java/robot/Ports.java:14]: But this goes by cartesian coordinate system convention where the top right quarter is first

**Review by orel-david — APPROVED** [review]
